README.md
# Password Generator for Atom

[![Build Status](https://travis-ci.org/maxkoryukov/atom-password-generator.svg?branch=master)](https://travis-ci.org/maxkoryukov/atom-password-generator)

Add generation of passwords in different strengths to the command palette:

* Generate short password: 10 alphanumeric characters
* Generate medium password: 15 alphanumeric characters
* Generate long password: 20 alphanumeric characters

Hotkeys:

`Ctrl + Alt + P` - generate medium password.

### Human confusing chars

There is a set of chars which could be interpreted wrong at the glance:

* `l` small L
* `I` capital i
* `|` vertical bar
* `1` one
* `0` zero
* `O` capital O letter
* `$` dollar sign

By default, the package avoids these letters, but you can enable using them
in the configuration at any time.

## CREDITS

Inspired by [Generate Password package for Sublime 3](https://packagecontrol.io/packages/Generate%20Password)

## LICENSE

**MIT**

But check updates: [LICENSE] on GitHub

  [LICENSE]: https://github.com/maxkoryukov/atom-password-generator/blob/master/LICENSE.md


CHANGELOG.md
# CHANGELOG

```
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################
```

> All notable changes to this project will be documented in this file.
>
> The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
> and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
>
> **This file MUST be filled only by maintainers, using messages from pull**
> **requests.**
>
> new release template
>
> ## [Unreleased] - yyyy-mm-dd
>
> ### Added
> ### Changed
> ### Fixed
> ### Removed

## 0.5.0 // 2020-04-26

- update dependencies

## 0.4.1 // 2017-11-28

* (gh #4) Usage of human-confusing chars is a configurable option
* Check that set of available chars is not empty
* Upgrade dev-dependencies

* (gh #6) FIXED: notify if the activeTextEditor was not found

## 0.3.0 // 2017-07-08

* (gh #3) Configurable length and chars for passwords (by [giovazz89](https://github.com/giovazz89))
* (188af71) add medium-length  password

Under the hood:

* A lot of refactoring
* enable eslint analysis

## 0.2 // 2016-04-04

* Javascript instead of Coffeescript

## 0.1.0  // 2015-12-05

First Release

* Every feature added (whaaaa???)
* Every bug fixed (whaaaa???)
* Travis CI integration


tests/runner.js
'use strict'


const createRunner = require('@atom/mocha-test-runner').createRunner
const mochaOptions = require('../.mocharc.js')

console.log(mochaOptions)

// optional options to customize the runner
/*
const extraOptions = {
	testSuffixes: ['-spec.js', '-spec.coffee']
}
*/


const optionalConfigurationFunction = function (/*mocha, {terminate}*/) {
	// If provided, atom-mocha-test-runner will pass the mocha instance
	// to this function, so you can do whatever you'd like to it.

	// The "terminate" function may be called to prevent the test suite from running. If it's called with an argument,
	// that will be used as the exit status of the main process.


}

module.exports = createRunner(
	mochaOptions
	//extraOptions
	, optionalConfigurationFunction
)


spec/.eslintrc.js
exports = module.exports = {
	'env': {
		'mocha': true,
	},
	'globals': {
		'expect': true
	},

	'rules': {
		'no-unused-expressions': 'off',   // because of `expect(x).to.be.true`
		'camelcase': 'off',  // don't care...
		'comma-dangle': ['error', {
			'arrays': 'only-multiline',
			'objects': 'only-multiline',
			'imports': 'only-multiline',
			'exports': 'only-multiline',
			'functions': 'ignore',
		}]
	}
}


tests/bootstrap.js
/*
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################
#
# Password generator for Atom
#
# Copyright (C) 2016-2020 Volebo <dev@volebo.net>
# Copyright (C) 2016-2020 Maksim Koryukov <maxkoryukov@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the MIT License, attached to this software package.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
# You should have received a copy of the MIT License along with this
# program. If not, see <https://opensource.org/licenses/MIT>.
#
# http://spdx.org/licenses/MIT
#
################################################################################
*/

'use strict'

const chai            = require('chai')
const tags            = require('mocha-tags')
const mh              = require('@volebo/mocha-helpers')

global.expect = chai.expect
global.tags = tags

global.filename2suitename = mh.filename2suitename


.mocharc.js
/*
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################
#
# Password generator for Atom
#
# Copyright (C) 2016-2020 Volebo <dev@volebo.net>
# Copyright (C) 2016-2020 Maksim Koryukov <maxkoryukov@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the MIT License, attached to this software package.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
# You should have received a copy of the MIT License along with this
# program. If not, see <https://opensource.org/licenses/MIT>.
#
# http://spdx.org/licenses/MIT
#
################################################################################
*/

'use strict'

module.exports = {
	reporter: 'list',
	// reporter: 'spec',

	recursive: true,
	checkLeaks: true,
	ui: 'bdd',
	require: 'tests/bootstrap',
	extension: ['test.js'],

	diff: true,
	//opts: './test/mocha.opts',
	package: './package.json',
	slow: 75,
	timeout: 2000,

	spec: [
		'src/**/*.test.js',
	]
}


nyc.config.js
/*
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################
#
# DESCRIPTION
#
# Copyright (C) 2016-2020 Volebo <dev@volebo.net>
# Copyright (C) 2016-2020 Maksim Koryukov <maxkoryukov@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the MIT License, attached to this software package.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
# You should have received a copy of the MIT License along with this
# program. If not, see <https://opensource.org/licenses/MIT>.
#
# http://spdx.org/licenses/MIT
#
################################################################################
*/

'use strict'

module.exports = {
	'all': true,

	// "sourceMap": false,
	// "instrument": false,

	'reporter': [
		'html',
		'lcov',
		'text',
		'text-summary',
		// 'json-summary',
	],

	'report-dir': './tmp/coverage',

	'include': [
		'src/**',
	],

	// 'exclude': [
	// 	'**/*.test.js',
	// 	'**/*.spec.js',
	// 	'tests/bootstrap.js',
	// ],
	// ["coverage/**","packages/*/test/**","test/**","test{,-*}.js","**/*{.,-}test.js","**/__tests__/**","**/{ava,babel,jest,nyc,rollup,webpack}.config.js"]]

	'excludeNodeModules': true
}


lib/password-generator.js
'use strict'

const Composite = require('atom').CompositeDisposable
const atom = global.atom

// TODO: gh #4 - remove human-confusing chars (l)
const lowercaseDic = 'abcdefghijkmnopqrstuvwxyz'
const uppercaseDic = 'ABCDEFGHJKLMNPQRSTUVWXYZ'
const numberDic    = '23456789'
const symbolsDic   = '!"#%&\'()*+,-./:;<=>?@[\\]^_`{}~'

const lowercaseDicConfusing = 'l'
const uppercaseDicConfusing = 'IO'
const numberDicConfusing    = '10'
const symbolsDicConfusing   = '$|'

const AtomPasswordGenerator = {

	config: {
		shortLength: {
			type:    'integer',
			default: 10,
			order:   0,
		},
		mediumLength: {
			type:    'integer',
			default: 15,
			order:   1,
		},
		longLength: {
			type:    'integer',
			default: 20,
			order:   2,
		},
		useConfusingChars: {
			type:    'boolean',
			default: false,
			order:   3,
		},
		useLowerCaseLetters: {
			type:    'boolean',
			default: true,
			order:   4,
		},
		useUpperCaseLetters: {
			type:    'boolean',
			default: true,
			order:   5,
		},
		useNumbers: {
			type:    'boolean',
			default: true,
			order:   6,
		},
		useSymbols: {
			type:    'boolean',
			default: false,
			order:   7,
		}
	},

	subscriptions: null,

	activate: function(_unused_state) {
		this.subscriptions = new Composite()

		const commands = {
			'password-generator:short': this.short.bind(this),
			'password-generator:medium': this.medium.bind(this),
			'password-generator:long': this.long.bind(this),
		}

		return this.subscriptions.add(
			atom.commands.add('atom-workspace', commands)
		)
	},

	deactivate: function() {
		return this.subscriptions.dispose()
	},

	getCharDomain: function(){
		let currentDictionary = ''
		const useConfusingChars = atom.config.get('atom-password-generator.useConfusingChars')

		if(atom.config.get('atom-password-generator.useLowerCaseLetters')) {
			currentDictionary += lowercaseDic
			if (useConfusingChars) {
				currentDictionary += lowercaseDicConfusing
			}
		}
		if(atom.config.get('atom-password-generator.useUpperCaseLetters')) {
			currentDictionary += uppercaseDic
			if (useConfusingChars) {
				currentDictionary += uppercaseDicConfusing
			}
		}
		if(atom.config.get('atom-password-generator.useNumbers')) {
			currentDictionary += numberDic
			if (useConfusingChars) {
				currentDictionary += numberDicConfusing
			}
		}
		if(atom.config.get('atom-password-generator.useSymbols')) {
			currentDictionary += symbolsDic
			if (useConfusingChars) {
				currentDictionary += symbolsDicConfusing
			}
		}

		return currentDictionary
	},

	getRandomString: function(dict, N){
		let s = ''
		const len = dict.length
		for (let i = 0; i < N; i++) {
			s += dict[Math.floor(Math.random() * len)]
		}
		return s
	},

	_gen: function(len) {
		let domain = this.getCharDomain()

		if (domain.length <= 0) {
			// TODO: gh #7 localize!
			const txtHead = 'No chars for password'
			const txtBody = 'All sets of chars are disabled in Setting. Plz, enable at least one!\n\n❤❤❤'

			atom.notifications.addWarning(
				txtHead, {
					description: txtBody,
					dismissable: false,
				})

			domain = '.'
		}

		const pwd = this.getRandomString(domain, len)
		const editor = atom.workspace.getActiveTextEditor()

		if (editor) {
			return editor.insertText(pwd)
		} else {
			// TODO: gh #7 localize!
			const txtHead = 'Active Text Editor missed'
			const txtBody = 'Plz, put the cursor into the editor pane and try again!\n\n❤❤❤'

			atom.notifications.addInfo(
				txtHead, {
					description: txtBody,
					dismissable: false,
				})
		}
	},

	short: function() {
		const len = atom.config.get('atom-password-generator.shortLength')
		return this._gen(len)
	},

	medium: function() {
		const len = atom.config.get('atom-password-generator.mediumLength')
		return this._gen(len)
	},

	long: function() {
		const len = atom.config.get('atom-password-generator.longLength')
		return this._gen(len)
	}
}

module.exports = AtomPasswordGenerator


.eslintrc.js
/*
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################
#
# Password generator for Atom
#
# Copyright (C) 2016-2020 Volebo <dev@volebo.net>
# Copyright (C) 2016-2020 Maksim Koryukov <maxkoryukov@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the MIT License, attached to this software package.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
# You should have received a copy of the MIT License along with this
# program. If not, see <https://opensource.org/licenses/MIT>.
#
# http://spdx.org/licenses/MIT
#
################################################################################
*/

const path = require('path')

exports = module.exports = {
	'root': true,

	'env': {
		node: true
		//'atom': true,
	},

	'globals': {
		'atom': true,
	},

	// 'parserOptions': {
	// 	'parser': 'babel-eslint'
	// },

	'extends': [
		// 'plugin:vue/essential',
		// 'volebo',
		path.join(__dirname, 'node_modules', 'eslint-config-volebo', 'index.js'),
	],

	'overrides': [
		{
			'files': [
				'**/*.test.js',
				'**/*.spec.js'
			],

			'env': {
				mocha: true
			},

			'globals': {
				'filename2suitename': true,
				'tags': true,
			},

			/*
			'rules': {
				'no-unused-expressions': 'off',   // because of `expect(x).to.be.true`
				'camelcase': 'off',  // don't care...
				'comma-dangle': ['error', {
					'arrays': 'only-multiline',
					'objects': 'only-multiline',
					'imports': 'only-multiline',
					'exports': 'only-multiline',
					'functions': 'ignore',
				}]
			},
			*/

		}
	],
}
